sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsGdsStore.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+
+import ai.rapids.cudf._
+import com.nvidia.spark.rapids.StorageTier.StorageTier
+import com.nvidia.spark.rapids.format.TableMeta
+
+import org.apache.spark.sql.rapids.RapidsDiskBlockManager
+import org.apache.spark.sql.types.DataType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/** A buffer store using GPUDirect Storage (GDS). */
+class RapidsGdsStore(
+    diskBlockManager: RapidsDiskBlockManager,
+    catalog: RapidsBufferCatalog = RapidsBufferCatalog.singleton)
+    extends RapidsBufferStore("gds", catalog) {
+  private[this] val sharedBufferFiles = new ConcurrentHashMap[RapidsBufferId, File]
+
+  override def createBuffer(
+      other: RapidsBuffer,
+      stream: Cuda.Stream): RapidsBufferBase = {
+    val otherBuffer = other.getMemoryBuffer
+    try {
+      val deviceBuffer = otherBuffer match {
+        case d: DeviceMemoryBuffer => d
+        case _ => throw new IllegalStateException("copying from buffer without device memory")
+      }
+      val id = other.id
+      val path = if (id.canShareDiskPaths) {
+        sharedBufferFiles.computeIfAbsent(id, _ => id.getDiskPath(diskBlockManager))
+      } else {
+        id.getDiskPath(diskBlockManager)
+      }
+      val fileOffset = if (id.canShareDiskPaths) {
```

Comment:
maybe a comment here would help with `copyDeviceBufferToFile`'s last argument. Without going to cudf to look there, I am not sure why we need to fork here.

Reply:
Done.